clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -119,6 +119,7 @@ public final class ConsumerCoordinator extends AbstractCoordinator {
     private Set<String> joinedSubscription;
     private MetadataSnapshot metadataSnapshot;
     private MetadataSnapshot assignmentSnapshot;
+    private boolean metadataUpdated;
```

Comment:
This optimization makes perfect sense but I am extremely paranoid that future changes will end up modifying the snapshot without remembering (or even knowing about) the need to update this boolean as well. Can you maybe encapsulate this in the `MetadataSnapshot` class or otherwise make sure it's impossible for this to get out of sync?

Reply:
Hmm, good point. I'm not sure how to encapsulate this inside `MetadataSnapshot` though, as the entire object is replaced when it's updated.

Would it be enough to place the swap in a private method, say:
```
private void updateAssignmentSnapshot() {
    assignmentSnapshot = metadataSnapshot;
    metadataUpdates = true;
}
```

Reply:
On reflection, I'm not convinced that this strategy will work, because there's already a shortcut on the `version` check, which means that the performance I'm seeing is when `assignmentSnapshot.version != metadataSnapshot.version`, so we definitely need to do the deep-equality check to determine if the new version of the metadata is actually different from the previous version.

I'm looking into some alternatives around whether we can represent this internal metadata differently, in a way that's more efficient to compare.

Reply:
I think I would personally need to see benchmarks (somehow) to be convinced that just changing it to a `LinkedHashSet` is sufficient as an optimization, or even just guaranteed not to be a regression. Speaking from experience, I once tried to optimize the cache in Streams for better range-scan performance which involved swapping out the underlying data structure for a `LinkedHashMap`. The consequences were pretty dire, with the `LinkedHashMap` having comparably terrible performance characteristics when scaling up the keyspace. Granted it's not the exact same scenario because that was a map and also the Concurrent variant which surely means the scaling characteristics are not going to be exactly the same. But it has made me incredibly suspicious of the `LinkedHashXXX` data structures for use cases like this.

All that said, if this is literally only used for an equality check and nothing else, why not just hash it or something like that?